### ISSUE TITLE
Top-level `unused_elements` are not ignored by `// ignore_for_file: type=lint`

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#4874]
Hello!  
Consider the following:

```dart
// ignore_for_file: type=lint

const _fn = 42;
```

This will warn that `_fn` is unused.  
My guess is that this is because some `unused_elements` diagnostics come-up as warnings, while some are infos (this such as this case).

**Comment by bwilkerson** [comment]
Correct. The `unused_element` diagnostic is a warning, not a lint, and hence not effected by the `type=lint`.

**Comment by rrousselGit** [comment]
It isn't a warning when affecting top-level variables, but an info

**Comment by rrousselGit** [comment]
Cf:

<img width="702" alt="Screenshot 2024-02-09 at 22 08 31" src="https://github.com/dart-lang/linter/assets/20165741/493ac686-b958-4885-a5b2-6020178f864d">

**Comment by bwilkerson** [comment]
You're right, it isn't a warning. It's one of a small number of diagnostics that are classified as a "hint", a classification that we're in the process of removing. I can't remember whether the plan is to make this a warning or to make it a lint.

@srawlins might be able to answer that question when he's back from some time off

**Comment by rrousselGit** [comment]
I'd prefer if this specific case became an info instead of a warning then.

The reasoning is, code-generators may want to silence all infos in generated code, but keep warnings.
And a code-generator may want to generate two classes, but a user may decide to use only one of them. In this case, there is an unused element, but it's not a big deal and would get tree shaken. Users shouldn't modify generated code, and adding APIs just for the sake of not generating this class would bloat the generator's API.

**Comment by srawlins** [comment]
The plan is to make it a warning.

It sounds like you want a way to configure the analyzer for a specific file. Something like "for all info-severity diagnostics reported here, just ignore them."

**Comment by rrousselGit** [comment]
Yes. that's what that `ignore_for_file: type=lint` does. It works fine with other lints like single vs double quote or missing trailling commas. It correctly ignores them.

Afaik, unused_element and unnecessary_cast are the only cases of "infos" at the moment that do not respect this global lint ignore.

**Comment by srawlins** [comment]
> > It sounds like you want a way to configure the analyzer for a specific file. Something like "for all info-severity diagnostics reported here, just ignore them."

> Yes. that's what that `ignore_for_file: type=lint` does.

That is not what that directive does. That directive ignores all diagnostics whose _type_ is "lint", not accounting for severity at all. "lint" is not a severity. "info" and "warning" and "error" are severities which can be configured in an analysis options file. The "type" that a diagnostic is cannot be configured.

**Comment by rrousselGit** [comment]
But there is a correlation between the severity and diagnostic type.
Most lints come up as infos. And non-lints typically don't.

Ultimately, the "type" of a diagnostic is an implementation detail. When `ignore: type=lint` was implemented, from a user PoV this feature is equal to "ignore infos".

**Comment by srawlins** [comment]
Maybe that is what you desire, but it is not what was designed nor implemented. Happy to keep this open though if you desire the behavior [here](https://github.com/dart-lang/linter/issues/4874#issuecomment-1942874351).

**Comment by bwilkerson** [comment]
> Ultimately, the "type" of a diagnostic is an implementation detail.

That doesn't match my mental model. I don't think it's about the implementation at all, I think it's about the nature of the diagnostic.

I define them this way:
- An _error_ is a diagnostic that is specified by the language specification as preventing the code from being run.
- A _warning_ is a diagnostic that won't prevent the code from being run, but is serious enough and has a low enough false positive rate that we have decided that it should be enabled for all users by default.
- A _lint_ is a diagnostic that doesn't qualify for either of the above and hence is only produced if the user has explicitly requested it.

The fact that we have chosen to group error codes by the type / kind of diagnostic is the implementation detail. We could have had a single `DiagnosticCode` class that defined all three kinds and it would have been a perfectly valid implementation.

The purpose of `// ignore_for_file: type=lint` is to allow users, or code generators, to disable all of the diagnostics that users are required to enable.

> But there is a correlation between the severity and diagnostic type.

That's true, and the correlation is strong. But the two are still separate concepts. The severity mostly impacts the way the diagnostics are displayed by tools (command-line and IDEs). Whether they ought to be separate concepts (or rather, whether we should model both concepts) is a separate conversation.

But, more to the point, I think what you're asking for (all questions of terminology aside) is that either
- some (at least one) case of `unused_element` be made a lint rather than a warning, or
- we add support for something like `// ignore_for_file: type=lint, warning`.

The latter isn't necessarily out of the question.

**Comment by rrousselGit** [comment]
> I define them this way:
> 
> * An _error_ is a diagnostic that is specified by the language specification as preventing the code from being run.
> * A _warning_ is a diagnostic that won't prevent the code from being run, but is serious enough and has a low enough false positive rate that we have decided that it should be enabled for all users by default.
> * A _lint_ is a diagnostic that doesn't qualify for either of the above and hence is only produced if the user has explicitly requested it.



Sure. But lint == info in that case.
A `ignore: type=lint` is never ignoring a warning or an error. It only ignores infos. Just that _some_ infos apparently bypasses this.


There may be a logical difference between "info" and "lint" in the SDK. But ultimately for users, I don't think that's correct.

That's why when we do `dart analyze` or `flutter analyze`, we have `--no-fatal-infos`, not `--no-fatal-lints`.
In that sense, I think it would be more appropriate to rename `ignore: type=lint` to `ignore: type=info`.

**Comment by bwilkerson** [comment]
> Sure. But lint == info in that case.

As we've defined it, there's no inherent reason why that has to be the case.

I actually thought we had chosen a severity of 'warning' for security-related lint codes. I'm not sure why we didn't, but I think it would have made sense to do so.

> There may be a logical difference between "info" and "lint" in the SDK. But ultimately for users, I don't think that's correct.

It might be the case that the distinction is not well understood by users. It might also be the case that the distinction isn't interesting to users. I can't definitively say in either case.

But in our current design the difference is real, and impacts what you're seeing, along with some other parts of our tooling. (I'm only trying to explain the tools here, not defend our design, in case that wasn't clear.)

For example, we allow users to change the severity of diagnostics, but we don't allow them to change the kind. If a user changes the severity of a lint from 'info' to 'warning', that doesn't cause the lint to be enabled by default: it's still a lint, not a warning. If kinds and severities were synonymous, then changing a diagnostic from an 'info'/'lint' to a 'warning' would make it be enabled by default. That isn't the case because we make a distinction between kind and severity.
